### PR TITLE
Interactive Login

### PR DESCRIPTION
**Pull request by kferrone**
## **User description**
This introduces an interactive login flow. This new flow introduced the need for a config class which can figure out how to get all configured prior to building the DuploClient. This also make the config class the owner of all of the global parameters.

### Try it out

Get a list of tenant names as yaml interactive mode.   
```sh
duploctl tenant list -q '[].AccountName' -o yaml --interactive
```

___

## **Type**
enhancement

___

## **Description**
- Introduced a new configuration management system (`DuploConfig`) to handle configurations and interactive login.
- Added global arguments for specifying home directory, cache directory, config file, and enabling interactive login mode.
- Refactored `DuploClient` to utilize `DuploConfig` for configuration, simplifying its implementation.
- Implemented a local server (`TokenServer`) to facilitate interactive login, capturing tokens from a web browser.
- Added `DuploExpiredCache` exception to handle scenarios where cached items are expired.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>args.py</strong><dd><code>Addition of New Global Arguments for Configuration and Interactive </code><br><code>Login</code></dd></summary>
<hr>
      
src/duplocloud/args.py

<li>Added new global arguments for home directory, cache directory, config <br>file, and interactive login mode.<br> <li> Updated help descriptions for new arguments.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-10408bee68a44ad34e4226ec91f20e3277d2d637fd8e17d3482fea817838ec96">+18/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>client.py</strong><dd><code>Refactor DuploClient to Use DuploConfig and Simplify Implementation</code></dd></summary>
<hr>
      
src/duplocloud/client.py

<li>Refactored DuploClient to use DuploConfig for configuration.<br> <li> Removed unused imports and simplified method implementations.<br> <li> Added <code>from_creds</code> static method to create a DuploClient from <br>credentials.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-ed921a2441e8df6f226f7024b875dea79afcd0d87e1326c37b19d232dba3f740">+47/-49</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>commander.py</strong><dd><code>Cleanup and Refactoring of Commander Module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/duplocloud/commander.py

<li>Removed unused imports and methods related to service loading.<br> <li> Renamed <code>load_service</code> to <code>load_resource</code> to reflect its purpose better.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-76bc85ae5e9281fbbae1c51d5775fc1663f2b323368722e8bf16a21578a9b9f1">+4/-21</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>config.py</strong><dd><code>New DuploConfig Class for Configuration Management and Interactive </code><br><code>Login</code></dd></summary>
<hr>
      
src/duplocloud/config.py

<li>Introduced DuploConfig class for handling configuration and <br>interactive login.<br> <li> Implemented methods for loading configuration from environment, <br>handling cache, and performing interactive login.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-6f9bd6f4b3f482a929bb442adfa0c549a49ec34d1c6fd4929958a88a70de7303">+206/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>errors.py</strong><dd><code>Introduction of DuploExpiredCache Exception</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/duplocloud/errors.py

<li>Added DuploExpiredCache exception for handling expired cache <br>scenarios.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-4519a687e27aee7d0417e452e8eb6d9a132924c122284200a806a4f68335b242">+14/-8</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>server.py</strong><dd><code>Implementation of Local Server for Interactive Login</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/duplocloud/server.py

<li>Implemented TokenServer and InteractiveLogin classes for handling <br>interactive login via a local server.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/35/files#diff-80ec0ae12053924aba45475fd6687a7bc7f1f77ddb5beff6937785a4f8f8ff6c">+59/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: src/duplocloud/config.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    Perform an interactive login to the specified host. Opens a temporary web browser to the login page and starts a local server to receive the token. When the user authorizes the request in the browser, the token is received and the server is shutdown.
+
+    Args:
+      host: The host to login to.
+
+    Returns:
+      The token as a string.
+    """
+    port = 56022 # this should be randomized. Will anyone catch this in the PR? 10 points if you do. 
+    page = f"{self.host}/app/user/verify-token?localAppName=duploctl&localPort={port}&isAdmin=true"
```
Comment (duplocloud-matt) [inline src/duplocloud/config.py:191]: This pass isAdmin=true.  Does this assume the user is always authenticating with an admin role?

File: src/duplocloud/config.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    
+    Perform an interactive login to the specified host. Opens a temporary web browser to the login page and starts a local server to receive the token. When the user authorizes the request in the browser, the token is received and the server is shutdown.
+
+    Args:
+      host: The host to login to.
+
+    Returns:
+      The token as a string.
+    """
+    port = 56022 # this should be randomized. Will anyone catch this in the PR? 10 points if you do.
```
[the remark this answers is not in the document]
Comment (kferrone) [reply src/duplocloud/config.py:190]: Points!!!
I do have a question here though. What are the params on randomizing? Between what and what?

File: src/duplocloud/config.py (hunk 1 above)
Comment (kferrone) [reply src/duplocloud/config.py:191]: I'm not actually sure the difference? I have not tried without isAdmin. Probably needs to be an argument. Like if isAdmin is false do I really get a temp token for a user?

File: src/duplocloud/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (duplocloud-matt) [reply src/duplocloud/config.py:190]: The dynamic and private port ranges is 49152-65535.

File: src/duplocloud/config.py (hunk 1 above)
Comment (duplocloud-matt) [reply src/duplocloud/config.py:191]: Looking at the duplocloud code, I believe isAdmin is significant.  It seems like if you don't pass isAdmin, you get a role scoped to a tenant.

File: src/duplocloud/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kferrone) [reply src/duplocloud/config.py:190]: I just found out making the port 0 will choose a random port.

**Review by duplocloud-matt — APPROVED** [review]

**Review by duplocloud-matt — APPROVED** [review]
Reviewed updated changes, this looks great!

**Pull request merged**